### ISSUE TITLE
Support running (via daemon) without the need to spawn Chrome with dev flags

### ISSUE THREAD

**Issue by DanTup** [dart-lang/webdev#403]
We briefly discussed this somewhere, but I don't think there's an open issue to track it. To support [some setups](https://code.visualstudio.com/docs/remote/remote-overview), we'd need to be able to run webdev on a remote machine (or in a container) without spawning the browser, then connect our own browser. This means webdev would have no connection to the Chrome debugger protocol.

As well as the general remote support, it seems likely [VS Code Online](https://techcrunch.com/2019/05/06/microsoft-launches-visual-studio-online-an-online-code-editor/) support would be enabled by this too.

**Comment by jakemac53** [comment]
Fwiw, webdev itself runs fine without connecting to chrome, its the `--debug` option specifically which won't work.

We are all in right now on the chrome debugger protocol for debugging, although there are other potential options that could eventually be explored, I doubt we will have time to look into this any time soon.

Not sure if we want to leave this issue open or not? @kevmoo @vsmenon

**Comment by vsmenon** [comment]
Doesn't this "just work" if you don't pass the `--debug` flag?  I.e., it'll still compile and serve in debug mode, but without the CDP support.

**Comment by DanTup** [comment]
Right now, we don't pass `--debug` anywhere at all. We just run `pub global run webdev daemon`.

To be clear, the issue here isn't just about support it, but about giving the daemon a way to actually run without launching a browser itself, and returning a URL to us so we can launch the browser ourselves (or rather, ask VS Code to).

**Comment by jakemac53** [comment]
Oh - I missed the `via daemon` part. I think we do always force `--debug` mode essentially with the daemon today. Maybe we should make that configurable?

**Comment by DanTup** [comment]
Yeah, that would help. Supporting only non-debug mode would still be a great win here (it'd be nice to have debug long term, but it's minor to being able to actually run anything, esp. given you can use browser tools :-))

I don't know whether it's best to make it opt-out or opt-in (and whether it's better to separate the browser launching from the debug flag, so you could run non-debug and still get browser, for ex.). Changing the default might upset existing users, but since it's preview it might be better to make the preferred API.

We should probably also consider how it'll look under `flutter run`. Currently for mobile we pass `--start-paused` when we want the debugger (and optionally `--observatory-port`).

**Comment by DanTup** [comment]
> Does vscode support webdev outside of flutter web right now?

Not currently (well not via F5.. the users can make tasks that will run it normally without the daemon). However I would like to do it soon (though maybe not this quarter) since I presume it's not massively different from Flutter web so shouldn't be a huge job?

(and this request applies equally there - since it'll allow us to support Angular/etc. in the same places :-))

**Comment by jakemac53** [comment]
> However I would like to do it soon (though maybe not this quarter) since I presume it's not massively different from Flutter web so shouldn't be a huge job?

In theory, there should be no difference at all :D. Webdev doesn't care about flutter in any specific way, other than it implements the flutter daemon protocol but that isn't really flutter specific either.

**Comment by grouma** [comment]
@DanTup is this issue still relevant?

**Comment by DanTup** [comment]
Hmm, I think so. As far as I can see, we still can't run using the daemon without Chrome being spawned.

In Flutter we can do this using the web-server device (instead of the Chrome device), but for non-Flutter web, we don't seem to have that.

However - the Flutter daemon had to evolve a little to properly support this though. Now it works like this:

- Ability to run without spawning a browser (for Flutter we use web-server device, but it could also be a flag)
- If the editor opts-in (with a flag), the daemon sends [app.exposeUrl](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/doc/daemon.md#appexposeurl) requests to the editor whenever it has a URL (for example localhost:xxxx) that would need to be exposed to the end users machine (this lets the editor set up tunnelling/port forwarding and return the "public" URL) - this happens for any backend servers like for the injected client backend/debug extension backend
- The daemon emits [`app.webLaunchUrl`](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/doc/daemon.md#appweblaunchurl) with the URL of the web app (so it can be opened by the client)

So it's a little more complicated than just a flag as described above. So I guess the first question is how important supporting non-Flutter web apps in cloud IDEs is. It's worth considering that right now, many of them don't support this even for Flutter web because of authentication issues (even when exposed, the Dart debug extension can't connect to the backend service because the backend service isn't really "public" but requires some auth). It's not clear if/when that will improve (though it is something I'd like to find solutions to - I think there's value in supporting these cloud IDEs).

**Comment by laurinneff** [comment]
For me this is an issue even without a cloud IDE as I'm running vscode in Flatpak, where it doesn't have access to the `google-chrome` executable.
